### PR TITLE
batches: maintain list state when deselecting a workspace

### LINKED ISSUE #40520: Closing workspace details clears workspace list state
On the execution page, when `x`-ing out of the workspace details, the state of the list is reset as we seem to drop all query params from the URL. This loses filters, for example.

### PR DESCRIPTION
**Pull request by adeola-ak**
Closes #40520

This will fix a bug that would do a page redirect and lose any state or search in the filter row if a workspace was deselected.

https://user-images.githubusercontent.com/67931373/189354015-ea76cebf-1759-49d4-b72b-46f471e6de8c.mp4

## Test plan
tested selecting and deselecting workspaces, as well as changing filter options

## App preview:

- [Web](https://sg-web-aa-maintain-workspace-state2.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-fntddnugav.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/web/src/enterprise/batches/batch-spec/execute/workspaces/WorkspacesFilterRow.tsx [hunk 1]
```diff
@@ -52,6 +52,9 @@ export const WorkspaceFilterRow: React.FunctionComponent<React.PropsWithChildren
         if (history.location.search !== searchParameters.toString()) {
             history.replace({ ...history.location, search: searchParameters.toString() })
         }
+
+        console.log(searchParameters.toString(), 'filter row')
+
```
Comment (eseliger) [inline client/web/src/enterprise/batches/batch-spec/execute/workspaces/WorkspacesFilterRow.tsx:57]:
```suggestion
```

**Review by eseliger — APPROVED** [review]
Nice, LGTM pending removal of the console.log statement.

**Review by Piszmog — APPROVED** [review]

File: client/web/src/enterprise/batches/batch-spec/execute/workspaces/ExecutionWorkspaces.tsx [hunk 2]
```diff
@@ -62,7 +62,9 @@ const MemoizedExecutionWorkspaces: React.FunctionComponent<
 }) {
     const history = useHistory()
 
-    const deselectWorkspace = useCallback(() => history.push(batchSpec.executionURL), [batchSpec.executionURL, history])
```
Comment (courier-new) [inline client/web/src/enterprise/batches/batch-spec/execute/workspaces/ExecutionWorkspaces.tsx:65]: Mentioned this to you over Slack already, but just for visibility -- looks like this and the other file were originally `history.push`, not `history.replace`, so we might want to retain that.

**Review by courier-new — APPROVED** [review]
🎉 😄

File: client/web/src/enterprise/batches/batch-spec/execute/workspaces/WorkspacesList.tsx [hunk 3]
```diff
@@ -44,7 +44,12 @@ export const WorkspacesList: React.FunctionComponent<React.PropsWithChildren<Wor
                         key={node.id}
                         workspace={node}
                         isSelected={node.id === selectedNode}
-                        onSelect={() => history.push(`${executionURL}/execution/workspaces/${node.id}`)}
+                        onSelect={() =>
+                            history.replace({
```
Comment (eseliger) [inline client/web/src/enterprise/batches/batch-spec/execute/workspaces/WorkspacesList.tsx:48]: was this left intentionally as `replace` instead of `push`?

**Pull request merged**